### PR TITLE
fix(agents): make a timed-out consult recoverable, and scope its history

### PR DESCRIPTION
**Pull request by 2witstudios**
## The symptom

`pagespace agents ask <agent> "<~2500 chars>"` returned:

> Request to agent … timed out — it may still be running … check the agent's conversation history before asking again.

The consult had **completed on the server**. Sending the same prompt again proved it: the agent opened with *"Same copy you sent me last round."* Credits spent, answer produced, answer unreachable — and the advice named a capability the CLI did not have.

## Root cause — four independent defects

**1. The deadline was unraisable by anyone.** `client.ts` computed `op.timeoutMsOverride ?? this.#timeoutMs`, so an operation's declared deadline *beat* an explicit client setting. `new PageSpaceClient({ timeoutMs: 600_000 })` still got `agents.ask`'s 120s; the CLI passed no `timeoutMs` at all. A fixed wall-clock cap is the wrong shape for a 20-step tool loop against an arbitrary model — but the real defect was that it was a **ceiling**, not that the number was wrong.

**2. Abandonment is one-sided, and silent.** The route never reads `request.signal`. A client timeout ends the *waiting*, never the work: `generateText` runs to completion, the credit hold settles, and both turns persist. The `TimeoutError` is a statement about the client, and it was being read as a statement about the server.

**3. The answer was unaddressable — not unreachable.** `activeConversationId = conversationId || createId()`, minted server-side and returned only in the 200 body. `createConversation` is called eagerly, so the row *exists* and `conversations.list`/`.read` can find it — those operations were already in the SDK and already served over MCP. They simply had no CLI verbs.

**4. Cross-conversation context bleed** (found while investigating). The no-`conversationId` path **wrote** to a fresh conversation while **reading** `getRecentPageMessagesForUser` — the caller's 10 most recent messages across *all* their conversations on that agent. Writes conversation-scoped, reads page-scoped. Consecutive one-shot consults silently saw each other, producing conversation rows each individually incomplete and collectively overlapping. It also broke parity with the internal `ask_agent` tool, which loads history only when given a `conversationId` — two entry points deliberately matched on their step budget were handing the same agent different context.

Verified `getRecentPageMessagesForUser` had exactly one non-test caller. Page chat (`getPageConversationMessages`) and `/v1/chat/completions` were never affected.

## Options considered

| | Why it loses |
|---|---|
| **Just raise the timeout** | Picks a bigger arbitrary number for something genuinely unbounded, and does nothing about reachability. Kept as one component, rejected as the fix. |
| **Async 202 + job id + poll** | Correct shape for unbounded work, but needs a job store, changes the contract for every consumer, and makes the MCP agent-to-agent story awkward (a tool that returns "pending"). Disproportionate to the defect. |
| **Idempotency keys** | Strictly stronger than never-retry — a repeat with the same key returns the stored result instead of re-executing. Wants a migration and a new retrieval operation, and fixes neither the CLI gap nor the bleed. The right *next* epic, not this one. |
| **Caller-addressed consultation** ✅ | The result was always retrievable; what was missing was an address and a verb. Reuses the conversation as the handle. No migration, no new operation. |

## What changed

- **`timeoutMsOverride` becomes a per-operation default**, not a ceiling: explicit caller > operation > fallback (`resolveTimeoutMs`). Every existing caller passes nothing, so nothing moves for them.
- **`--timeout <seconds>`** and **`PAGESPACE_TIMEOUT_MS`**. The env var also covers `pagespace mcp`, which builds the same client — so this is not a CLI-only fix.
- **`newConversationId`** on `agents.ask`: a caller-minted address, known before the request is sent. Deliberately a **separate field** from `conversationId` — folding "create here" into it would have turned that path's unknown-id 404 (identical to someone else's id, on purpose) into an existence oracle for every id. A collision is refused with **409**, never appended to.
- **`pagespace conversations list|read`** over the existing SDK operations. No new SDK surface; the registry stays the single source of truth for all three facades.
- **The timeout message** now names the retrieval command, the conversation id, and `--timeout`.
- **History is conversation-scoped in both branches.** The cross-conversation reader is removed — it had no other caller, and leaving it would fail knip.

## On the trade-offs

- The **409 is a mild existence oracle** for one caller-supplied id. Accepted knowingly: ids are cuid2, the caller supplied it, and the alternative — appending this turn to whatever conversation already lives there, possibly another user's — is categorically worse. Documented at the call site.
- **Removing `getRecentPageMessagesForUser` discards its owner-scoping hardening** and two integration tests. That predicate was the right containment for the wrong read: scoping a leak to one user makes it not a leak, but left every consult seeing that user's other conversations. With no caller, there is nothing left to scope. The tests are replaced by emptiness assertions in `new-conversation-history.test.ts`.
- **Consults are no longer implicitly stateful.** Anyone relying on the accidental memory must now pass a `conversationId`. That is the parity the internal tool already had.

## Tests

Existing tests were treated as documentation of intent and changed deliberately:

- `client.test.ts`'s override block asserted the operation beat "the client default" while passing an **explicit** 1000ms — it could not distinguish the two readings, which is how the bug survived. Replaced with a four-case precedence matrix, and `makeClient` can now omit `timeoutMs` at all.
- `recent-history-ordering.test.ts` → `new-conversation-history.test.ts`. #1769 was right about ordering and wrong about scope; there is no window left to order. Emptiness is asserted **by value** against a 15-message fixture that stays in the database — proving it against an empty stub would pass just as happily if the route resumed reading tomorrow.
- `conversation-continuity.test.ts`'s "fallback asks for THIS caller's messages" is superseded and says why.
- **Mutation-checked**: restoring the original `op.timeoutMsOverride ?? explicit` precedence turns 4 tests red.

## Verification

SDK: **816 tests pass** locally. `scaffold.test.ts` fails on a missing built `dist` — a worktree environment gap, not this change.

apps/web and packages/cli tests were **not run locally**: this worktree has no `node_modules` and `@tailwindcss/postcss` is absent even from the root install, and CLI tests import the built `dist`. Both need an install plus two package builds, which the concurrent-agent load rule rules out. **CI is the gate for those two surfaces.**

Closes the defect reported against `agents.ask`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013rsKjDroSs1w3PoFePGYFU

## Summary by CodeRabbit

* **New Features**
  * Added conversation listing and reading commands with JSON and readable output.
  * Added `--timeout` support with environment configuration and validation.
  * Timed-out agent requests now provide an exact recovery command.
  * New conversations can use a caller-specified ID for reliable recovery and follow-up.

* **Behavior Changes**
  * New consultations start with empty history.
  * Conversation IDs are validated, including conflicts and invalid combinations.
  * Timeout precedence is applied consistently across the CLI and SDK.
  * Updated API compatibility information for caller-specified conversation IDs.

### CODE REVIEW

File: apps/web/src/app/api/ai/page-agents/consult/route.ts [hunk 1]
```diff
@@ -306,9 +324,47 @@ export async function POST(request: Request) {
...
+    // identically so an id-guessing caller learns nothing, and accepting
+    // unknown ids there as "create" would have converted it into an existence
+    // oracle for every conversation id. A caller-supplied NEW id that already
+    // exists is refused with 409 — which does reveal that this one id is
+    // taken, accepted knowingly: ids are cuid2, the caller supplied the id
+    // itself, and the alternative (appending this turn to whatever
+    // conversation already lives at that address, possibly another user's) is
+    // categorically worse.
+    if (newConversationId) {
+      const existing = await conversationRepository.getConversation(newConversationId);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/consult/route.ts]: Correct, and this was the exact failure the 409 was meant to prevent — fixed in 6687249b1.

The check was also redundant: `createConversation` already inserts with `onConflictDoNothing().returning()` and returns `'exists'` when the row was not written, with a comment stating explicitly that a concurrent first-write winning the race is `'exists'` too. The route was discarding that outcome with `.catch(() => {})` and then asking the same question again, unsafely.

The write is now the reservation: a caller-supplied `newConversationId` must come back `'created'`, and `'exists'` or `'message_owner_conflict'` is a 409. Ignoring the outcome stays correct for a server-minted id and for the `conversationId` path (which has already authorised what it names) — both now have tests so neither starts 409ing.

The old test would have passed against the racy code, since it mocked the pre-check. It now drives `createConversation`s

File: apps/web/src/app/api/ai/page-agents/consult/route.ts [hunk 2]
```diff
@@ -219,6 +219,24 @@ export async function POST(request: Request) {
...
+    // contradiction, not something to resolve by precedence — silently
+    // preferring one would write the turn somewhere the caller did not name.
+    if (conversationId && newConversationId) {
+      return NextResponse.json(
+        { error: 'Pass conversationId to continue a conversation or newConversationId to start one, not both' },
+        { status: 400 }
+      );
+    }
+
+    if (newConversationId !== undefined && (typeof newConversationId !== 'string' || newConversationId.length === 0)) {
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/consult/route.ts]: Agreed — fixed in 6687249b1. `newConversationId` is now validated against `^[a-z][a-z0-9]{1,31}$` rather than merely being non-empty, so a caller-addressed conversation cannot persist an address nothing else in the system would mint.

Fixtures moved to CUID2-shaped ids, and the invalid-input case is now a table covering an empty string, a uuid, an uppercase id, a leading digit, a path fragment and an oversized string.

File: apps/web/src/app/api/ai/page-agents/consult/route.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/consult/route.ts]: Addressed in 6687249b1 — same fix as the P1 thread above. The reservation is now the insert itself (`createConversation`'s `'created'` outcome), not a preceding SELECT, so the read-then-write window is gone.

File: packages/cli/src/commands/agents.ts [hunk 3]
```diff
@@ -139,47 +152,87 @@ function scanValueFlags(args: readonly string[], flags: readonly string[]): Flag
...
+    ctx.stdout.write(`${result.response}\n\n(conversationId: ${result.conversationId})\n`);
     return EXIT_SUCCESS;
-  }
-  ctx.stdout.write(`${result.response}\n\n(conversationId: ${result.conversationId})\n`);
-  return EXIT_SUCCESS;
-};
+  };
+}
+
+export const agentsAskHandler: CommandHandler = createAgentsAskHandler({ newConversationId: () => randomUUID() });
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/cli/src/commands/agents.ts]: Fixed in 6687249b1 — the CLI now mints the CUID2 shape (`^[a-z][a-z0-9]{1,31}$`) rather than a uuid, which the route would now refuse outright.

One deliberate deviation: it does not pull in `@paralleldrive/cuid2`. `@pagespace/cli` is published, so a runtime dependency is one every user installs, and what the route requires is the *format*, not a particular generator. It renders 128 CSPRNG bits in base36 as a single BigInt — not a per-character `byte % 36`, which would bias the low digits harmlessly at this size but visibly to anyone who copied it.

The assertion checks the shape, since the format is now part of the contract rather than an implementation detail.

File: packages/cli/src/commands/conversations.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * message, which would read as "the agent said nothing" when in fact it did
+ * a great deal.
+ */
+export function renderMessageParts(message: ConversationMessage): string {
+  const rendered = message.parts
+    .map((part) => {
+      if (typeof part.text === 'string' && part.text.length > 0) return part.text;
+      if (part.type.startsWith('tool-')) return `[tool: ${part.toolName ?? part.type.slice('tool-'.length)}]`;
+      if (part.type === 'file') return `[file: ${part.filename ?? part.mediaType ?? 'attachment'}]`;
+      return '';
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/cli/src/commands/conversations.ts]: Good catch — the code contradicted its own doc comment, which claimed every non-text part rendered as a labelled placeholder. Fixed in 0148fc0d7: unknown parts are now named, so a message made only of `data-*` parts no longer renders blank. That mattered more than it looks, because the `parts` union grows server-side — the failure gets quieter over time rather than louder.

One deviation from the suggestion: `step-start` still returns `''`. It is a structural marker with no content, so labelling it would put noise in every transcript rather than information. That exception is now asserted, so it stays a decision rather than becoming an accident.

Coverage added for a standalone `data-*` part, an unknown future type, a message made only of unknown parts, and the `step-start` exception.

File: packages/cli/src/argv/parse.ts [hunk 5]
```diff
@@ -94,7 +107,17 @@ export function parseArgv(argv: readonly string[]): ParseResult {
...
+      else if (flagName === '--key') key = value;
+      else {
+        // Seconds in, milliseconds out. Finite and > 0 only: 0 and negatives
+        // would mean "already expired", and NaN would silently become the
+        // default the caller was trying to override.
+        const seconds = Number(value);
+        if (!Number.isFinite(seconds) || seconds <= 0) {
+          return { kind: 'usage-error', message: 'Flag --timeout requires a positive number of seconds.' };
+        }
+        timeoutMs = Math.round(seconds * 1000);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/cli/src/argv/parse.ts]: Reopening: this was **not** addressed in a620efb..955dab9 — those commits never touched this logic. The finding was correct and I verified it against the source before fixing.

```
--timeout 0.0001         -> Math.round(0.1) = 0ms
--timeout 1e308          -> 1e308 * 1000    = Infinity
PAGESPACE_TIMEOUT_MS=0.4 -> Math.round(0.4) = 0ms
```

All three are positive and finite going in, which is why the pre-conversion check missed them. Coming out, 0ms aborts every request instantly, and Node's `setTimeout` silently substitutes 1ms for anything past 2^31-1, so an absurdly *long* timeout aborts instantly too.

It matters more than "minor" here: an explicit `timeoutMs` outranks every operation's own default, so the caller asking to wait longer is precisely the one whose requests stop working — and the `TimeoutError` is indistinguishable from a server hang-up.

Fixed by validating the converted value in both entry points against a shared `isUsableTimeoutMs` (>= 1ms, <= setTimeout's ceiling), with boundary tests on both.

File: packages/cli/src/argv/parse.ts [hunk 6]
```diff
@@ -94,7 +111,26 @@ export function parseArgv(argv: readonly string[]): ParseResult {
...
+        // unusable values come from: 0.0001 is a positive finite number of
+        // seconds that rounds to 0ms, and 1e308 is finite until multiplied by
+        // 1000. Both yield a client that aborts every request immediately —
+        // and since an explicit timeout outranks each operation's own default,
+        // the caller asking to wait longer is the one who breaks.
+        const requestedMs = Math.round(Number(value) * 1000);
+        if (!isUsableTimeoutMs(requestedMs)) {
+          return {
+            kind: 'usage-error',
+            message: `Flag --timeout requires a number of seconds between 0.001 and ${Math.floor(MAX_TIMEOUT_MS / 1000)}.`,
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/cli/src/argv/parse.ts]: Correct on both boundaries — fixed.

`0.0005` rounds up to 1ms and `2147483.5` converts to 2147483500ms, so the seconds range in the message described neither what the code enforces nor what a caller at the edge observes. Since milliseconds are what is actually validated, the message now quotes milliseconds rather than tightening the check to match a seconds range that rounding makes fictional either way.

Boundary cases are now asserted on the accepting side too (`0.0005` -> 1ms, `2147483.5` -> 2147483500ms), not just the rejecting side, plus a test that the message names the unit and the ceiling it checks.

**Comment by 2witstudios** [comment]
All posted findings are addressed and their threads resolved (7/7), but neither reviewer has seen the current head:

- Codex last reviewed `a620efbe4` — commit 1 of 9. It has **not** seen `fa422c658`, which fixes the P1 TOCTOU it reported.
- CodeRabbit last reviewed `6162417a8` — it has not seen `d6bc274f6`.

Requesting a fresh pass on the final state rather than assuming the fixes are right.

@codex review
@coderabbitai review

File: packages/cli/src/commands/agents.ts [hunk 7]
```diff
@@ -139,47 +157,111 @@ function scanValueFlags(args: readonly string[], flags: readonly string[]): Flag
...
+    const minted = continuing === undefined ? deps.newConversationId() : undefined;
+
+    let result: Awaited<ReturnType<PageSpaceClient['agents']['ask']>>;
+    try {
+      result = await ctx.sdk.agents.ask({
+        agentId,
+        question,
+        context: scanned.values.get('--context'),
+        conversationId: continuing,
+        newConversationId: minted,
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/cli/src/commands/agents.ts]: Real, and fixed in bdb38116e — though not by version gating, for a reason worth stating.

Gating on the observed server version cannot work at this call site. The SDK learns the server's version from the first 2xx response, and for a bare `pagespace agents ask` the timing-out request **is** the first request — nothing came back, so no `X-PageSpace-API-Version` was ever observed. A gate would therefore read "unknown" exactly when the message is produced, degrading to the cautious wording every time and losing the precision on servers that *do* honour the field.

What was actually wrong was the message asserting the address as fact. It now offers it as the address the CLI **asked for**, names the version that honours it, and keeps the always-correct `conversations list` fallback beside it — true against either server:

```
Retrieve the answer with:
  pagespace conversations read <agent> <id>

That is the address this CLI asked the server to use. A server older than API 1.1.0 ignores the request and
picks its own, so if the read finds nothing, list what actually exists:
  pagespace conversations list <agent>
```

Separately, `API_CONTRACT_VERSION` is bumped 1.0.0 -> 1.1.0, which was owed regardless: its own doc says it is bumped by PRs that change the contract, and this one adds a request field. `MIN_SERVER_API_VERSION` stays at 1.0.0 deliberately — a client built against 1.1.0 still works against a 1.0.0 server for every operation, it just does not get to choose the address. Raising the floor would reject older servers outright for *all* operations, which is far more than this warrants.

File: apps/web/src/app/api/ai/page-agents/consult/route.ts [hunk 8]
```diff
@@ -306,38 +333,93 @@ export async function POST(request: Request) {
...
-    await conversationRepository.createConversation(activeConversationId, userId, agentId).catch(() => {});
-
-    // Both branches feed the target agent's model context, so both exclude 'streaming'
-    // placeholders (empty, mid-flight rows) — see Server Stream Durability epic PR 2. The
-    // fallback branch was also missing the isActive filter the conversationId branch already
-    // had (pre-existing gap, called out in the PR 2 board's reader inventory).
+    // invisible to list_conversations.
+    const createOutcome = await conversationRepository
+      .createConversation(activeConversationId, userId, agentId)
+      .catch(() => 'error' as const);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/page-agents/consult/route.ts]: Correct — fixed in bdb38116e. A rejected `createConversation` (database outage, deadlock) was folded into the same branch as `'exists'`, so a repository failure reported 409 "that id already exists" for an id nothing had ever taken, with retry guidance that could not work.

The `.catch()` is gone. Thrown failures now reach the outer handler as a retryable 500, and 409 is reserved for the repository's explicit non-created outcomes (`'exists'`, `'message_owner_conflict'`).

File: packages/cli/src/argv/__tests__/parse.test.ts [hunk 9]
```diff
@@ -217,3 +217,101 @@ describe('parseArgv', () => {
...
+  /**
+   * The bound is on the CONVERTED milliseconds, and rounding means the
+   * seconds-side boundary is not a round number: `0.0005` rounds up into
+   * range and `2147483.5` stays in range. The message therefore quotes
+   * milliseconds — quoting seconds would describe a range that is neither
+   * what is enforced nor what a caller at the edge observes.
+   */
+  it.each([
+    ['a fractional second that rounds up into range', '0.0005', 1],
+    ['a fractional second at the top of the range', '2147483.5', 2_147_483_500],
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/cli/src/argv/__tests__/parse.test.ts]: Right — `2147483.5`s is `2147483500`ms, 147ms short of the real ceiling, so a regression in the bound would have slipped past it. Fixed in bdb38116e with `2147483.647` (accepted, exactly `MAX_TIMEOUT_MS`) and `2147483.648` (rejected).
